pkg/operator: fix errors

alaypatel07 opened the pull request:
1) util.go was failing on empty string
2) cluster member controller adds to etcd membership only if MemberReady
3) Observer etcd isPendingReady returns true if certs container exists
and the pod is not crashlooping

alaypatel07 committed:
pkg/operator: fix errors

1) util.go was failing on empty string
2) cluster member controller adds only if MemberReady
3) Observer etcd isPendingReady returns true if certs container exists
and the pod is not crashlooping

alaypatel07 committed:
manifests: add permissions to list endpoints for setup-env

hexfusion commented on pkg/operator/configobservation/etcd/observe_etcd.go:
```diff
@@ -190,10 +191,30 @@ func ObservePendingClusterMembers(genericListers configobserver.Listers, recorde
+	} else {
+		if pod.Status.InitContainerStatuses[1].State.Terminated != nil && pod.Status.InitContainerStatuses[1].State.Terminated.ExitCode == 0 {
+			if pod.Status.ContainerStatuses[0].State.Terminated != nil && pod.Status.ContainerStatuses[0].RestartCount > 0 {
+				klog.Info("isPendingRead: restart count >0, Remove state")
```
This makes me feel that we could become wedged if an initcontainer[0] for unknown reasons has a restart. Why do we care if discovery has a restart but then exit 0? How does Pod recover if `pod.Status.ContainerStatuses[0].RestartCount > 0`

alaypatel07 replied:
Good point. 

For instance, the discovery container could fail during bootstrap because dns in not set up. But in that case, we loop through to wait for dns to not really failing exiting with 0. Looking at discovery code, is there a way to make reasonable assumptions that discovery will never exit or exit 0? 

Also, if discovery indeed exits say a non 0 number, it will be restarted by kube until it exits 0 before certs container starts. If 
`pod.Status.InitContainerStatuses[1].State.Terminated.ExitCode == 0`, my understanding is that it would be safe to assume discovery exit code is 0.

and lastly my assumption was that after the restart controller takes the pod out from remove state, it RestartCount will be set to 0, so it will still recover. I can see now how this assumption might be wrong.

alaypatel07 replied:
I think we need to wire something up for multiple scenarios:

Scenario 1: Bootstrapping, member is in `PendingUnknown` State
Scenario 2: Restart, member is in `Remove` State

I suggest the following: 

1. If the member is in PendingUnknown State, and if we see two containers terminated with exit code 0, return `true`
2. If the member is in Remove State and if we see two containers terminated return `true`

The fact that after `Remove`, restart controller will kick in and restart the pod, will make sure that the init containers are run in order, once both of them terminated successfully, our code will return `Ready`, and the member will be added again. Does that cover most edge cases? @hexfusion

alaypatel07 commented on manifests/0000_12_etcd-operator_00_staticpod-rolebinding.yaml:
```diff
@@ -7,6 +7,9 @@ rules:
 - apiGroups: ["machineconfiguration.openshift.io"]
   resources: ["machineconfigpools", "machineconfigs"]
   verbs: ["get", "list"]
+- apiGroups: ["v1"]
```
typo s/"v1"/""

I'll update the PR shortly

alaypatel07 committed:
pkg/*/observe_etcd.go: update isPendingReady

isPendingReady returns true only when etcd-member container is waiting
and 2 init containers have successfully exit. This will take care of
1) Bootstrap: Containers will be waiting for PodInitializing state
2) Restart: Restart controller will remove the pod, making it
in PodInitializing state

hexfusion commented on pkg/operator/clustermembercontroller/clustermembercontroller.go:
```diff
@@ -180,7 +180,7 @@ func (c *ClusterMemberController) sync() error {
 		// Pending MemberAdd: here we have observed the static pod having add dependencies filled ok to scale Cluster API.
-		if c.IsStatus("pending", p.Name, ceoapi.MemberAdd) {
+		if c.IsStatus("pending", p.Name, ceoapi.MemberReady) {
 			if err := c.etcdMemberAdd([]string{fmt.Sprintf("https://%s:2380", peerFQDN)}); err != nil {
```
I still would like to have a distinction between a member who is in the que for addition and one that is dependency complete. Because the observation for `MemberReady` is done on the success of the init containers the Pod would technically still be `MemberUnknown` if that did not happen. I think for right now your change is fine but we might need to revisit.

alaypatel07 replied:
hrm makes sense, I think it would be good to something that acts as a selector, that pick one up from all the members that are dependency ready and makes it `MemberReady`.

hexfusion reviewed (approved):
/lgtm
